s4/sync.py

```diff
@@ -219,80 +170,110 @@ def get_sync_states(self, keys=None):
-                deferred_calls[key] = DeferredFunction(
-                    self.create_client, self.client_1, self.client_2, key, state_2.local_timestamp
+                resolutions[key] = Resolution(
+                    'CREATE', self.client_1, self.client_2, key, state_2.local_timestamp
                 )
 
             elif (
                 state_2.state == SyncState.DELETED and
                 state_1.state == SyncState.CREATED and
                 state_1.remote_timestamp == state_2.remote_timestamp
             ):
-                deferred_calls[key] = DeferredFunction(
-                    self.create_client, self.client_2, self.client_1, key, state_1.local_timestamp
+                resolutions[key] = Resolution(
+                    'CREATE', self.client_2, self.client_1, key, state_1.local_timestamp
                 )
 
             else:
                 unhandled_events[key] = (state_1, state_2)
 
-            self.logger.debug('Action=%s', deferred_calls.get(key))
+            self.logger.debug('Action=%s', resolutions.get(key))
 
-        return deferred_calls, unhandled_events
+        return resolutions, unhandled_events
 
-    def run_deferred_calls(self, deferred_calls):
+    def run_resolutions(self, resolutions, dry_run=False):
         # call everything once we know we can handle all of it
-        self.logger.debug('There are %s total deferred calls', len(deferred_calls))
+        self.logger.debug('There are %s total deferred calls', len(resolutions))
         success = []
         try:
-            for key in sorted(deferred_calls.keys()):
-                deferred_function = deferred_calls[key]
+            for key in sorted(resolutions.keys()):
+                resolution = resolutions[key]
+                if resolution.action == 'UPDATE':
+                    self.logger.info(
+                        colored.yellow('Updating %s (%s => %s)'),
+                        resolution.key,
+                        resolution.from_client.get_uri(),
+                        resolution.to_client.get_uri()
+                    )
+                    deferred_function = self.move_client
+                elif resolution.action == 'CREATE':
+                    self.logger.info(
+                        colored.green('Creating %s (%s => %s)'),
+                        resolution.key,
+                        resolution.from_client.get_uri(),
+                        resolution.to_client.get_uri()
+                    )
+                    deferred_function = self.move_client
+                elif resolution.action == 'DELETE':
+                    self.logger.info(
+                        colored.red('Deleting %s on %s'),
+                        resolution.key,
+                        resolution.to_client.get_uri()
+                    )
+                    deferred_function = self.delete_client
+                else:
+                    raise ValueError('Unknown resolution', resolution)
+
+                if dry_run:
```

Comment:
Try simplify this. It should also be clearer that operations are not actually being performed